### ISSUE TITLE
Siamese network implementation 

### ISSUE THREAD

**Issue by shaih82** [benanne/Lasagne#168]
hey all,
does anyone can advice on how to implement siamese network using lasagne?

I'm missing contrastive loss

i've looked at the CAFFE implementation of this,
and i can't understand where the "margin" param came from
is it a constant?

Computes the contrastive loss $ E = \frac{1}{2N} \sum\limits_{n=1}^N \left(y\right) d + \left(1-y\right) \max \left(margin-d, 0\right) $ where $ d = \left| \left| a_n - b_n \right| \right|_2^2 $.
![image](https://cloud.githubusercontent.com/assets/110537/6656733/ba4b5ad0-cb3b-11e4-8832-8520396b2871.png)

**Comment by benanne** [comment]
Hi,

currently we don't have an implementation of this in Lasagne, but it should be easy enough to just implement the formula verbatim.

To make the network "siamese", I would suggest building a single 'copy' of the network and glueing images together in pairs so they form a single minibatch. So instead of having two batches `batch_a` and `batch_b` of shape `(N, ...)`, for the 'left' and 'right' examples respectively, you would have a single batch of shape `(2N, ...)` where the examples with even indices are the left examples, and the examples with odd indices are the right examples.

Then you can simply reshape the output of the network and retrieve the learned representations for the left and right examples as `a = output[0::2]` and `b = output[1::2]`. 

So the loss can be implemented in Theano as:
```
a = output[0::2]
b = output[1::2]
d = T.sum((a - b)**2, axis=1)  # this is now a vector with distances for each pair of examples (a, b)
loss = T.mean(y * d + (1 - y) * T.maximum(margin - d, 0))
```

Here, `margin` is a hyperparameter that controls how far dissimilar points should be pushed apart. The underlying idea is as follows: if the representations `a` and `b` of two points are further than `sqrt(margin)` apart, the loss will be zero, so they will not be pushed apart further.

**Comment by shaih82** [comment]
great thanks :),
i actually managed to handle the inputs,
i'v just created one copy of the network, and pass the left and the right
batches separately
like so:

net = get_net(input_shape, output_dim, batch_size)
> a = T.ftensor4('a')
> b = T.ftensor4('b')
> y = T.ivector('y')
> z1 = net.get_output(a)
> z2 = net.get_output(b)
> d = T.sum((z1-z2)**2,axis=1)



and if were at it,
1. the y labels should be [0,1] or [-1,1] or is it even matter?
2. does the net last layer should be softmax or just dense layer with a
long feature vector?

**Comment by benanne** [comment]
Your approach will probably be fine, but if your minibatch size is fairly small, gluing the two sides together will be a bit faster because it increases parallelism in the computations. So it's a small optimization, whether it's worth it will depend on how big your network is.

1. 1 and 0 makes the most sense, I think.
2. It will work best when the layer is linear (i.e. no nonlinearity) because you are essentially learning an embedding into a Euclidean space.

This type of questions is more appropriate for our mailing list though: https://groups.google.com/forum/#!forum/lasagne-users
The GitHub issues are only for development, bug reports and feature requests.

**Comment by shaih82** [comment]
ok, got it,
the labels should be 0,1
1 - for same class
0 - for not-same class

**Comment by HKUSTmji** [comment]
@benanne your information indeed help me~ Thanks. But I am wondering why you said: "It will work best when the layer is linear (i.e. no nonlinearity) because you are essentially learning an embedding into a Euclidean space." 
What's the embedding concept come from? Because I only got results about 'word embedding' from google.
Why does the linear layer make more sense?

**Comment by f0k** [comment]
> Where does the embedding concept come from?

You're defining a network that produces a vector for each input example. The network is then trained to increase or decrease the euclidean distance (called `d` in the opening post) for pairs of such vectors (called `a` and `b` in the opening post) depending on whether they belong to the same class or not.

Since the network produces vectors, and the vectors are compared by Euclidean distance, the network learns an embedding of objects into a Euclidean space where you can (hopefully) compare objects in a meaningful way -- similar to how a word embedding produces vectors for words to make it easier to compare them in a meaningful way.

> Why does the linear layer make more sense?

Because it allows the network to use the full, untransformed output space. If you place a nonlinearity in front of the output, the vectors will be squashed (for a sigmoid function) or crippled (for a rectifier), and the Euclidean distance doesn't really fit any more.

**Comment by HKUSTmji** [comment]
@f0k Thanks for your explanation. Any popular paper about this 'embedding' concept?

**Comment by f0k** [comment]
Maybe read this one: http://yann.lecun.com/exdb/publis/pdf/chopra-05.pdf, or this one: http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf -- in general, look for "metric learning".

**Comment by HKUSTmji** [comment]
@f0k Thank you. I understand this metric learning lets the Euclidian distance on the mapped manifold describe the semantic similarity of the inputs. The transform is highly non-linear anyway, why we care about the linearity of the last NN layer?
Let's take a 1D toy example, input/output pair x/y like this: y=1 if x>0 else 0. 
What we want is a transform f(x), so that the f(x1)-f(x2) can reflect the similarity of the inputs x1 & x2.
One possible transform could be a sigmoid function, which is non-linear. But the sigmoid(x1)-sigmoid(x2) indeed works and is meaningful?

**Comment by f0k** [comment]
> But the sigmoid(x1)-sigmoid(x2) indeed works and is meaningful?

No, this could become negative. You need to come up with something different. And depending on how you do it, the distance measure may become more difficult to interpret. For sigmoids, cross-entropy is probably easier to justify than Euclidean distance. Also think about the learning dynamics: For sigmoids with Euclidean distance, for a pair of points, it's easier to move the one that is closer to the center of the hypercube (the gradient is larger), creating some possibly unwanted asymmetry.

> The transform is highly non-linear anyway, why we care about the linearity of the last NN layer?

True. If the previous to last layer is sigmoidal, adding a linear layer doesn't really make the Euclidean distance more justifiable. If you have rectifiers in the hidden layers, it may be different.
